File: Algorithms/Greedy-algorithms/Ford-Fulkerson.py
""" Ford Fulkerson algorithm in python:"""



class Graph:
    def __init__(self, graph):
        self.graph = graph
        self.ROW = len(graph)

    """ using BFS as a searching algorithm:"""
    def bfsSearch(self, s, t, parent):
        visited = [False] * self.ROW
        queue = []

        #append source: 
        queue.append(s)
        visited[s] = True

        while queue:
            u = queue.pop(0)
            for ind, val in enumerate(self.graph[u]):
                if visited[ind] == False and val > 0:
                    queue.append(ind)
                    visited[ind] = True
                    parent[ind] = u
        return True if visited[t] else False
    
    def FordFulkerson(self, source, sink):
        parent = [-1] * self.ROW
        maxFlow = 0

        while self.bfsSearch(source, sink, parent):
            print(parent)
            pathFlow = float("inf")
            s = sink
            while(s != source):
                pathFlow = min(pathFlow, self.graph[parent[s]][s])
                s = parent[s]
            """ adding the path flows:"""
            maxFlow += pathFlow

            """ updating the residual values of the edges:"""
            v = sink
            while(v != source):
                u = parent[v]
                self.graph[u][v] -= pathFlow
                self.graph[v][u] += pathFlow
                v = parent[v]
        return maxFlow

if __name__ == "__main__":
    graph = [[0, 8, 0, 0, 3, 0],
            [0, 0, 9, 0, 0, 0],
            [0, 0, 0, 0, 7, 2],
            [0, 0, 0, 0, 0, 5],
            [0, 0, 7, 4, 0, 0],
            [0, 0, 0, 0, 0, 0]]
    g = Graph(graph)

    source = 0
    sink = 5

    print("Max flow %d " % g.FordFulkerson(source, sink))

File: Algorithms/Greedy-algorithms/Ford-Fulkerson.cpp
/* Ford Fulkerson implementation in cpp:*/

#include "iostream"
#include "bits/stdc++.h"

#define V 6

using namespace std;

bool bfs(int residualGraph[V][V], int s, int t, int * parent){
    bool visited[V];
    memset(visited, 0, sizeof(visited));

    queue<int> q;
    q.push(s);
    visited[s] = true;
    parent[s] = -1;

    while(!q.empty()){
        int u = q.front();
        q.pop();

        for (int v = 0; v < V; v ++){
            if (visited[v] == false && residualGraph[u][v] > 0){
                q.push(v);
                parent[v] = u;
                visited[v] = true;
            }
        }

    }
    return (visited[t] == true);
}

/* applying ford Fulkerson algorithm:*/

int fordFulkerson(int graph[V][V], int s, int t){
    int u, v;

    int residualGraph[V][V];
    for (u = 0; u < V; u ++){
        for(v = 0; v < V ;v ++){
            residualGraph[u][v] = graph[u][v];
        }
    }

    int parent[V];
    int maxFlow = 0;

    /* updating the residual values in the graph:*/
    while(bfs(residualGraph, s, t, parent)){
        int pathFlow = INT_MAX;
        for (v = t; v != s; v = parent[v]){
            u = parent[v];
            pathFlow = min(pathFlow, residualGraph[u][v]);
        }

        for (v = t; v != s; v = parent[v]){
            u = parent[v];
            residualGraph[u][v] -= pathFlow;
            residualGraph[v][u] += pathFlow;
        }

        /* adding the path flows:*/
        maxFlow += pathFlow;
    }

    return maxFlow;
    
}

int main(){
    int graph[V][V] = {{0, 8, 0, 0, 3, 0},
                        {0, 0, 9, 0, 0, 0},
                        {0, 0, 0, 0, 7, 2},
                        {0, 0, 0, 0, 0, 5},
                        {0, 0, 7, 4, 0, 0},
                        {0, 0, 0, 0, 0, 0}};
    cout << "Max flow is -> " << fordFulkerson(graph, 0, 5) << endl;
    return 0;
}